apps/remote_control/lib/lexical/remote_control/code_intelligence/ast/aliases.ex

```diff
@@ -0,0 +1,214 @@
+       ) do
+    # This is for unexpanded aliases that leading with `__MODULE__`
+    # like: alias __MODULE__.{MyChild, MyOtherChild}
+    {:ok, parent_modules} = fetch_aliases_of_current_module(quoted, position[:line])
+    unexpanded = unexpanded_alias_mapping(parent_modules, children_ast)
+    {ast, Map.merge(acc, unexpanded)}
+  end
+
+  defp do_collect({:defmodule, _, [{:__aliases__, _, aliases} | _]} = ast, acc, quoted) do
+    # This is for implicit aliases, like the `State` module in a GenServer implementation Module
+    # or some nested defmodules
+    {ast, Map.merge(acc, implicit_alias_mapping(quoted, aliases))}
+  end
+
+  defp do_collect(ast, acc, _quoted) do
+    {ast, acc}
+  end
+
+  defp implicit_alias_mapping(quoted, aliases) do
+    path_to_defmodule =
+      Macro.path(
+        quoted,
+        &match?({:defmodule, _, [{:__aliases__, _, ^aliases} | _]}, &1)
+      )
+
+    modules =
+      path_to_defmodule
+      |> Enum.reduce([], fn
+        {:defmodule, _, [{:__aliases__, _, aliases} | _]}, acc ->
+          acc ++ [Module.concat(aliases)]
+
+        _, acc ->
+          acc
+      end)
+
+    modules_length = length(modules)
+
+    modules
+    |> Enum.with_index(fn module, index ->
+      ancestors_to_current_module =
+        modules |> Enum.slice(index, modules_length) |> Enum.reverse() |> Module.concat()
+
+      {module, ancestors_to_current_module}
+    end)
+    |> Map.new()
+  end
+
+  defp cursor_current_module_mapping(quoted) do
+    path_to_cursor = Macro.path(quoted, &match?({:__cursor__, _, _}, &1))
+
+    modules =
+      path_to_cursor
+      |> Enum.reduce([], fn
+        {:defmodule, _, [{:__aliases__, _, aliases} | _]}, acc ->
+          [Module.concat(aliases ++ acc)]
+
+        _, acc ->
+          acc
+      end)
+
+    if modules != [] do
+      [module] = modules
+      %{:__MODULE__ => module}
+    else
+      %{}
+    end
+  end
+
+  defp unexpanded_alias_mapping(parent_modules, children_ast) do
+    for {:__aliases__, _, [child]} <- children_ast, into: %{} do
+      {Module.concat([child]), Module.concat(parent_modules ++ [child])}
+    end
+  end
+
+  defp fetch_aliases_of_current_module(quoted, alias_line) do
+    path =
+      Macro.path(quoted, fn
+        {:defmodule, [line: line], _} when line < alias_line -> true
+        _ -> false
+      end)
+
+    if path do
+      [defs | _] = path
+      {:defmodule, _, [{:__aliases__, _, aliases} | _]} = defs
+      {:ok, aliases}
+    else
+      Logger.info("cannot find the module definition of __MODULE__ in #{alias_line}")
+      :error
+    end
+  end
+
+  defp full(quoted, original) do
+    [head | tail] = original
+
+    with true <- current_module_as_atom?(head),
+         {:__MODULE__, position, _} = head,
+         {:ok, aliases} <- fetch_aliases_of_current_module(quoted, position[:line]) do
+      Module.concat(aliases ++ tail)
+    else
+      _ -> Module.concat(original)
+    end
+  end
+
+  defp current_module_as_atom?(module) do
```

Comment:
this function is misnamed. It is checking for a three tuple where the first element is `:__MODULE__`, yet uses the word `as` which implies some conversion is taking place. 
Suggest: `current_module_alias?` Though I'm not sure why you're matching like that at all, given line 203. 
why not (in the `full` function)
```elixir
with {:__MODULE__, metadata, _} <- head, 
     {:ok, line} <- Keyword.fetch(metadata, :line),
    {:ok, aliases} <- fetch_current_module_aliases(quoted, line) do 
  Module.concat(aliases ++ tail)
else _ ->
  Module.concat(original)
end
```

Reply:
we can't just use `{:__MODULE__, metadata, _} <- head` to do the match, that will trigger a matching exception when the head is like "HeadModule"

Reply:
> to do the match, that will trigger a matching exception
That's not how with statements work. If it fails to match, it'll take the else, not raise a `MatchError` what _you're_ doing with the `=` on the line below will raise an exception though.

Reply:
example:
```elixir
with {:__MODULE__, metadata, _} <- :hi do
  IO.puts "here"
else _ ->
   IO.puts "there"
end
```